src/argilla/server/contexts/datasets.py

```diff
@@ -509,83 +519,190 @@ async def _validate_suggestion(
-                    raise ValueError(f"Provided user_id: {response.user_id!r} is not a valid user id")
-
-                validate_response_values(dataset, values=response.values, status=response.status)
-
-                record.responses.append(
-                    Response(
-                        values=jsonable_encoder(response.values),
-                        status=response.status,
-                        user_id=response.user_id,
-                    )
-                )
-
-        if record_create.suggestions:
-            for suggestion in record_create.suggestions:
-                try:
-                    questions = await _validate_suggestion(db, suggestion, questions=questions)
-                except ValueError as e:
-                    raise ValueError(f"Provided suggestion for record at position {record_i} is not valid: {e}") from e
-
-                record.suggestions.append(
-                    Suggestion(
-                        type=suggestion.type,
-                        score=suggestion.score,
-                        value=suggestion.value,
-                        agent=suggestion.agent,
-                        question_id=suggestion.question_id,
-                    )
-                )
-
-        if record_create.metadata:
-            try:
-                metadata_properties = await _validate_metadata(
-                    db,
-                    dataset=dataset,
-                    metadata=record_create.metadata,
-                    metadata_properties=metadata_properties,
-                )
-            except ValueError as e:
-                raise ValueError(f"Provided metadata for record at position {record_i} is not valid: {e}") from e
-
         records.append(record)
+        all_vectors.extend(record_vectors)
 
     async with db.begin_nested():
         db.add_all(records)
         await db.flush(records)
+        # TODO: We need to remove these asynchronicity requirements somehow
         for record in records:
             await record.awaitable_attrs.responses
+        for vector in all_vectors:
+            await vector.awaitable_attrs.vector_settings
+        # TODO: Review and unify in a single method
         await search_engine.index_records(dataset, records)
+        if all_vectors:
+            await search_engine.set_records_vectors(dataset, all_vectors)
 
     await db.commit()
 
 
+async def validate_record_metadata(
+    db: "AsyncSession",
+    dataset: Dataset,
+    record_idx: int,
+    metadata: Dict[str, Any],
+    cache: Dict[str, MetadataProperty] = {},
+):
+    """Validate metadata for a record."""
+    if not metadata:
+        return
+
+    try:
+        await _validate_metadata(db, dataset=dataset, metadata=metadata, metadata_properties=cache)
+    except ValueError as e:
+        raise ValueError(f"Provided metadata for record at position {record_idx} is not valid: {e}") from e
+
+
+async def create_record_vectors(
```

Comment:
I believe we should use `build` (or similar) for these functions instead of `create`. `create` should be used when we insert a new row/rows into the database.

These functions should be "private" also.

```suggestion
async def _build_record_vectors(
```

Reply:
Done